Added bits functions

Izaakwltn opened the pull request:
Added a few functions to `coalton-library/bits`.

`dpb` and `ldb` mimic their Common Lisp counterparts, following the same argument order.

`print-bits` is a handy debugging tool for printing bit types to a specified number of bits.

stylewarning commented:
Please keep the original argument order where the bytespec comes first. I actually think that's better for currying. 

I would also get rid of the bytespecs entirely. Just pass them as curried args.

Izaakwltn commented:
> Please keep the original argument order where the bytespec comes first. I actually think that's better for currying.
> 
> I would also get rid of the bytespecs entirely. Just pass them as curried args.

Changed the order, removed Bytespec

stylewarning commented on library/bits.lisp:
```diff
@@ -36,7 +40,25 @@
-           (Integer -> :int -> :int))))
+           (Integer -> :int -> :int)))
+
+  (declare dpb (Bits :a => UFix -> UFix -> UFix -> :a -> :a))
```
should the type of newbyte be :a ?

stylewarning commented on library/bits.lisp:
```diff
@@ -36,7 +40,25 @@
+  (define (print-bits width x)
+    "Prints an integer to a binary string, preserving leading zeros to a given width."
+    (lisp String (x width)
+               (cl:format cl:nil (cl:format cl:nil "~~~a,'0b" width) x))))
```
indentation

stylewarning replied:
you don't need a double format. use

    ~v,'0B

stylewarning commented on library/bits.lisp:
```diff
@@ -36,7 +40,25 @@
+      (cl:ldb (cl:byte size position) bitstring)))
+
+  (declare print-bits (Bits :a => UFix -> :a -> String))
+  (define (print-bits width x)
```
i expect "print" to be side-effectful. Maybe "bits->string"?

stylewarning commented on library/bits.lisp:
```diff
@@ -36,7 +39,25 @@
+      (cl:ldb (cl:byte size position) bitstring)))
+
+  (declare bits->string (Bits :a => UFix -> :a -> String))
+  (define (bits->string width x)
```
since x is a type class constrained type, it's not guaranteed to be a compatible Lisp value with FORMAT. you need to convert it to an integer for this to work, or extract the bits one by one 

i suggest either fixing this or removing it

Izaakwltn replied:
Sounds good, I removed it. 

It would fit better in a potential future Coalton formatting library anyways.
